clarity/contracts/pool-token/fwp-wstx-alex-tranched-120.clar

```diff
@@ -0,0 +1,303 @@
+        (ok { dx: stx, dy: alex })
+    )
+)
+
+(define-public (claim-and-stake (reward-cycle uint))
+  (let 
+    (      
+      ;; claim all that's available to claim for the reward-cycle
+      (claimed (and (> (as-contract (get-user-id)) u0) (is-ok (as-contract (claim-staking-reward reward-cycle)))))      
+      (alex-balance (unwrap! (contract-call? .age000-governance-token get-balance-fixed (as-contract tx-sender)) ERR-GET-BALANCE-FIXED-FAIL))
+      (principal-balance (unwrap! (contract-call? .fwp-wstx-alex-50-50-v1-01 get-balance-fixed (as-contract tx-sender)) ERR-GET-BALANCE-FIXED-FAIL))
+      (bounty (var-get bounty-in-fixed))
+      (current-cycle (unwrap! (get-reward-cycle block-height) ERR-STAKING-NOT-AVAILABLE))
+    )
+    (asserts! (>= block-height (var-get start-block)) ERR-NOT-ACTIVATED)
+    (asserts! (> current-cycle reward-cycle) ERR-REWARD-CYCLE-NOT-COMPLETED)
+    (asserts! (> alex-balance bounty) ERR-INSUFFICIENT-BALANCE)
+    (asserts! (>= (var-get end-cycle) current-cycle) ERR-STAKING-NOT-AVAILABLE)
+
+    (let 
+      (
+        (sender tx-sender)
+        (cycles-to-stake (if (>= (var-get end-cycle) (+ current-cycle u32)) u32 (- (var-get end-cycle) current-cycle)))
+      )
+      (and (> principal-balance u0) (> cycles-to-stake u0) (as-contract (try! (stake-tokens principal-balance cycles-to-stake))))
+      ;; this needs to be done carefully
+      ;; (alex-claimed (and (> (as-contract (get-alex-user-id)) u0) (is-ok (as-contract (claim-alex-staking-reward reward-cycle)))))
+      ;; (and (> cycles-to-stake u0) (as-contract (try! (stake-alex-tokens (/ (- alex-balance bounty) u2) cycles-to-stake))))
+      (and (> bounty u0) (as-contract (try! (contract-call? .age000-governance-token transfer-fixed bounty tx-sender sender none))))
+    
+      (ok true)
+    )
+  )
+)
+
+(define-private (distribute-iter (recipient principal) (prior (response { balance: uint, sum: uint } uint)))
+  (let 
+    (      
+      (prior-unwrapped (try! prior))
+      (shares (div-down (mul-down (get balance prior-unwrapped) (get-user-balance-or-default recipient)) (var-get total-balance)))
+    )
+    (as-contract (try! (contract-call? .age000-governance-token transfer-fixed shares tx-sender recipient none)))
+    (ok { balance: (get balance prior-unwrapped), sum: (+ (get sum prior-unwrapped) shares) })
+  )
+)
+
+(define-read-only (is-cycle-batch-processed (cycle uint) (batch uint))
+  (default-to
+    false
+    (map-get? processed-batches { cycle: cycle, batch: batch })
+  )
+)
+
+(define-public (distribute (cycle uint) (batch uint) (recipients (list 200 principal)))
+	(let 
+    (
+      (alex-balance (unwrap! (contract-call? .age000-governance-token get-balance-fixed (as-contract tx-sender)) ERR-GET-BALANCE-FIXED-FAIL))
+    )  
+		(asserts! (or (is-ok (check-is-owner)) (is-ok (check-is-approved))) ERR-NOT-AUTHORIZED)
+    (asserts! (is-eq (is-cycle-batch-processed cycle batch) false) ERR-ALREADY-PROCESSED)
```

Comment:
is there a way to prevent duplicate entries?

e.g. the bot sending in it data with some incorrect offset? a user appeared in two batches?

should we instead of guarding against batch, we guard with 

address + cycle? and ignore the ones already processed?

Maybe the cost (both runtime and development) cost out weights the benefits?

Reply:
It's tricky, but I may be able to implement a guard. Let me see.

Reply:
@zhigang1992 I added a few guards to ensure:
- users cannot add/reduce positions until distribution for the previous cycle is completed (`distributed`)
- a user cannot receive distribution twice (`user-distributed`)